llvm/lib/Target/AArch64/AArch64SchedNeoverseN3.td

```diff
@@ -0,0 +1,2359 @@
+                                         "^LD1S?B_[HSD]$",
+                                         "^LD1S?H_[SD]$",
+                                         "^LD1S?W_D$" )>;
+
+// Contiguous load broadcast, scalar + imm
+def : InstRW<[N3Write_6c_1L], (instregex "^LD1R[BHWD]_IMM$",
+                                         "^LD1RSW_IMM$",
+                                         "^LD1RS?B_[HSD]_IMM$",
+                                         "^LD1RS?H_[SD]_IMM$",
+                                         "^LD1RS?W_D_IMM$",
+                                         "^LD1RQ_[BHWD]_IMM$")>;
+
+// Contiguous load broadcast, scalar + scalar
+def : InstRW<[N3Write_6c_1L], (instregex "^LD1RQ_[BHWD]$")>;
+
+// Non temporal load, scalar + imm
+def : InstRW<[N3Write_6c_1L], (instregex "^LDNT1[BHWD]_ZRI$")>;
+
+// Non temporal load, scalar + scalar
+def : InstRW<[N3Write_6c_1L], (instregex "^LDNT1[BHWD]_ZRR$")>;
+
+// Non temporal gather load, vector + scalar 32-bit element size
+def : InstRW<[N3Write_7c_4L], (instregex "^LDNT1[BHW]_ZZR_S$",
+                                         "^LDNT1S[BH]_ZZR_S$")>;
+
+// Non temporal gather load, vector + scalar 64-bit element size
+def : InstRW<[N3Write_6c_3L], (instregex "^LDNT1S?[BHW]_ZZR_D$")>;
+def : InstRW<[N3Write_6c_3L], (instrs LDNT1D_ZZR_D)>;
+
+// Contiguous first faulting load, scalar + scalar
+def : InstRW<[N3Write_6c_1L], (instregex "^LDFF1[BHWD]$",
+                                         "^LDFF1S?B_[HSD]$",
+                                         "^LDFF1S?H_[SD]$",
+                                         "^LDFF1S?W_D$")>;
+
+// Contiguous non faulting load, scalar + imm
+def : InstRW<[N3Write_6c_1L], (instregex "^LDNF1[BHWD]_IMM$",
+                                         "^LDNF1S?B_[HSD]_IMM$",
+                                         "^LDNF1S?H_[SD]_IMM$",
+                                         "^LDNF1S?W_D_IMM$")>;
+
+// Contiguous Load two structures to two vectors, scalar + imm
+def : InstRW<[N3Write_8c_1V_1L], (instregex "^LD2[BHWD]_IMM$")>;
+
+// Contiguous Load two structures to two vectors, scalar + scalar
+def : InstRW<[N3Write_8c_1V_1L], (instregex "^LD2[BHWD]$")>;
+
+// Contiguous Load three structures to three vectors, scalar + imm
+def : InstRW<[N3Write_8c_3V_4L], (instregex "^LD3D_IMM$")>;
+
+// Contiguous Load three structures to three vectors, scalar + imm
+def : InstRW<[N3Write_10c_6V_9L], (instregex "^LD3[BHW]_IMM$")>;
+
+// Contiguous Load three structures to three vectors, scalar + scalar
+def : InstRW<[N3Write_9c_3V_4L_6I], (instregex "^LD3D$")>;
+
+// Contiguous Load three structures to three vectors, scalar + scalar
+def : InstRW<[N3Write_11c_6V_9L_12I], (instregex "^LD3[BHW]$")>;
+
+// Contiguous Load four structures to four vectors, scalar + imm
+def : InstRW<[N3Write_8c_4V_6L], (instregex "^LD4D_IMM$")>;
+
+// Contiguous Load four structures to four vectors, scalar + imm
+def : InstRW<[N3Write_12c_5V_7L], (instregex "^LD4[BHW]_IMM$")>;
```

Comment:
Where do these various (5V, 7L) numbers come from?

Reply:
Here "5V" means 5 uops for V and "7L" means 7 uops for L. How I get these is described in the comment above explaining how I get "2V".

Reply:
So kind of random guessing? Would 4L + 8V work OK too?

Reply:
(mV, nL) where m <= 5, n <= 7 should work.

I choose those numbers to match the throughput assuming all utilized ports are fully utilized.

Reply:
There may be some level of approximation in the SWOG values. If there are already a breakdown in the V2 scheduling model then that would probably be a good guide to follow: https://github.com/llvm/llvm-project/blob/6023d17e6b6624913b85fe9d2b5d79ae681e5764/llvm/lib/Target/AArch64/AArch64SchedNeoverseV2.td#L2673

Reply:
In the V2 SWOG:

Instruction Group | AArch64 Instructions | Exec Latency | Execution Throughput | Utilized Pipelines
-- | -- | -- | -- | --
Contiguous Load four structures to four vectors, scalar + imm  | LD4B, ... | 9 | 1/2 | V, L

To match the throughput, the write should be `V2Write_9cyc_mL_nV` where `(m == 6, n <= 8)` or `(m <= 6, n == 8)`. To determine m and n, we need to either know more microarchitecture details or conduct some experiments in a real machine.

Reply:
That is also the case for N3. Since I have neither access to those microarchitecture details nor a real N3 machine, there is nothing more I can do to improve the write here.